pkg/core/storage/boltdb_store.go

```diff
@@ -0,0 +1,129 @@
+package storage
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"os"
+
+	"github.com/CityOfZion/neo-go/config"
+	"github.com/boltdb/bolt"
+	"github.com/syndtr/goleveldb/leveldb/util"
+)
+
+var (
+	Bucket = []byte("DB") // Bucket name
+)
+
+// BoltDB is the storage implementation for storing and retrieving
+// blockchain data.
+type BoltDBStore struct {
+	db *bolt.DB
+}
+
+// BoltDbBatch simple batch implementation to satisfy the Store interface.
+type BoltDbBatch struct {
+	mem map[*[]byte][]byte
+}
+
+// Len implements the Batch interface.
+func (b *BoltDbBatch) Len() int {
+	return len(b.mem)
+}
+
+// Put implements the Batch interface.
+func (b *BoltDbBatch) Put(k, v []byte) {
+	b.mem[&k] = v
+}
+
+// NewRBoltDBBatch returns a new ready to use BoltDbBatch.
+func NewRBoltDBBatch() *BoltDbBatch {
+	return &BoltDbBatch{
+		mem: make(map[*[]byte][]byte),
+	}
+}
+
+// NewBoltDBStore returns a new ready to use BoltDB storage with created bucket.
+func NewBoltDBStore(ctx context.Context, cfg config.BoltDBOptions) (*BoltDBStore, error) {
+	fileName := cfg.FilePath
+	db, err := bolt.Open(fileName, os.FileMode(0666), nil)
+	if err != nil {
+		return nil, err
+	}
+	var bucket *bolt.Bucket
+	err = db.Update(func(tx *bolt.Tx) error {
+		bucket, err = tx.CreateBucketIfNotExists(Bucket)
+		if err != nil {
+			return fmt.Errorf("could not create root bucket: %v", err)
+		}
+		return nil
+	})
+
+	// graceful shutdown
+	go func() {
+		<-ctx.Done()
+		db.Close()
+	}()
+
+	return &BoltDBStore{db: db}, nil
+}
+
+// Put implements the Store interface.
+func (s *BoltDBStore) Put(key, value []byte) error {
+	err := s.db.Update(func(tx *bolt.Tx) error {
+		b := tx.Bucket(Bucket)
+		err := b.Put(key, value)
+		return err
+	})
+	return err
+}
+
+// Get implements the Store interface.
+func (s *BoltDBStore) Get(key []byte) ([]byte, error) {
+	var v []byte = nil
+	err := s.db.View(func(tx *bolt.Tx) error {
+		b := tx.Bucket(Bucket)
+		v = b.Get(key)
+		return nil
+	})
+	if v == nil {
+		return nil, ErrKeyNotFound
+	}
+	return v, err
+}
+
+// PutBatch implements the Store interface.
+func (s *BoltDBStore) PutBatch(batch Batch) error {
+	err := s.db.Batch(func(tx *bolt.Tx) error {
+		b := tx.Bucket(Bucket)
+		for k, v := range batch.(*BoltDbBatch).mem {
+			err := b.Put(*k, v)
+			if err != nil {
+				return err
+			}
+		}
+		return nil
+	})
+	return err
+}
+
+// Seek implements the Store interface.
+func (s *BoltDBStore) Seek(key []byte, f func(k, v []byte)) {
+	err := s.db.View(func(tx *bolt.Tx) error {
+		c := tx.Bucket(Bucket).Cursor()
+		prefix := util.BytesPrefix(key)
+		for k, v := c.Seek(prefix.Start); k != nil && bytes.Compare(k, prefix.Limit) <= 0; k, v = c.Next() {
+			f(k, v)
+		}
+		return nil
+	})
+	if err != nil {
+		fmt.Println("error while executing seek in boltDB")
+	}
+}
+
+// Batch implements the Batch interface and returns a boltdb
+// compatible Batch.
+func (s *BoltDBStore) Batch() Batch {
+	return NewRBoltDBBatch()
```

Comment:
Maybe we don't need `NewRBoltDBBatch()` at all then?

Reply:
misspell =/